pandas/tests/dtypes/test_cast.py

```diff
@@ -28,10 +31,727 @@
+    fill_dtype = np.dtype(string_dtype)
+
+    if dtype not in (bool, str, object) and boxed and box_dtype is None:
+        pytest.xfail('does not upcast to object')
+    if (not issubclass(dtype.type, (np.datetime64, np.timedelta64))
+            and not boxed):
+        pytest.xfail('does not return correct missing value marker')
+
+    # create array of given dtype
+    fill_value = np.array(['abc'], dtype=fill_dtype)
+
+    if not boxed:
+        fill_value = fill_value[0]
+    elif box_dtype == object:
+        fill_value = fill_value.astype(box_dtype)
+
+    # filling anything with string casts to object
+    expected_dtype = np.dtype(object)
+    expected_na_value = np.nan
+
+    result_dtype, result_na_value = maybe_promote(dtype, fill_value)
+    assert result_dtype == expected_dtype
+    assert expected_na_value is np.nan and result_na_value is np.nan
+
+
+@pytest.mark.parametrize('boxed, box_dtype', [
+    (True, None),    # fill_value wrapped in array with auto-dtype
+    (True, object),  # fill_value wrapped in array with object dtype
+    (False, None)    # fill_value directly
+])
+def test_maybe_promote_object_with_any(object_dtype, any_numpy_dtype,
+                                       boxed, box_dtype):
+    dtype = np.dtype(object_dtype)
+    fill_dtype = np.dtype(any_numpy_dtype)
+
+    if (issubclass(fill_dtype.type, (np.datetime64, np.timedelta64))
+            and boxed and box_dtype is None):
+        pytest.xfail('falsely returns iNaT as missing_value')
+    elif not boxed:
+        pytest.xfail('does not return correct missing value marker')
+
+    # create array of given dtype; casts "1" to correct dtype
+    fill_value = np.array([1], dtype=fill_dtype)
+
+    if not boxed:
+        fill_value = fill_value[0]
+    elif box_dtype == object:
+        fill_value = fill_value.astype(box_dtype)
+
+    # filling object with anything remains object
+    expected_dtype = np.dtype(object)
+    expected_na_value = np.nan
+
+    result_dtype, result_na_value = maybe_promote(dtype, fill_value)
+    assert result_dtype == expected_dtype
+    assert expected_na_value is np.nan and result_na_value is np.nan
+
+
+@pytest.mark.parametrize('boxed, box_dtype', [
+    (True, None),    # fill_value wrapped in array with auto-dtype
+    (True, object),  # fill_value wrapped in array with object dtype
+    (False, None)    # fill_value directly
+])
+def test_maybe_promote_any_with_object(any_numpy_dtype, object_dtype,
+                                       boxed, box_dtype):
+    dtype = np.dtype(any_numpy_dtype)
+
+    if (not issubclass(dtype.type, (np.datetime64, np.timedelta64))
+            and not boxed):
+        pytest.xfail('does not return correct missing value marker')
+
+    # create array of object dtype
+    fill_value = np.array([{'example', 'of', 'non-castable', 'object'}])
+
+    if not boxed:
+        fill_value = fill_value[0]
+    elif box_dtype == object:
+        fill_value = fill_value.astype(box_dtype)
+
```

Comment:
this is too big of an add here, make a new file. you move the existing tests  to 
pandas/tests/dtypes/cast/test_cast.py then add this. but needs to be split up this is too massive.

Reply:
The existing tests are already in `pandas/tests/dtypes/cast/test_cast.py`. Do you want me to make a new file `pandas/tests/dtypes/cast/test_promote.py`?

Reply:
existing are in pandas/tests/dtypes/test_cast.py, so make a subdir 'cast' and move the original, then add new in test_promote in that subdir